src/main.rs

```diff
@@ -33,98 +34,110 @@ mod template;
 type Feature<'a> = &'a str;
 type PackageName<'a> = &'a str;
 type RootFeature<'a> = (PackageName<'a>, Feature<'a>);
+type Result<T> = std::result::Result<T, Error>;
 
 const VERSION_ATTRIBUTE_NAME: &str = "cargo2nixVersion";
 
 fn main() {
     let args: Vec<String> = std::env::args().collect();
     let args: Vec<&str> = args.iter().map(AsRef::as_ref).collect();
+    if let Err(err) = try_main(&args) {
+        eprint!(colorify!(red_bold: "error: "));
+        eprintln!("{}", err);
+        std::process::exit(1);
+    }
+}
 
+fn try_main(args: &[&str]) -> Result<()> {
     match &args[1..] {
         ["--stdout"] | ["-s"] => generate_cargo_nix(io::stdout().lock()),
         ["--file"] | ["-f"] => write_to_file("Cargo.nix"),
         ["--file", file] | ["-f", file] => write_to_file(file),
         ["--help"] | ["-h"] => print_help(),
-        ["--version"] | ["-v"] => println!("{}", version()),
+        ["--version"] | ["-v"] => {
+            println!("{}", version());
+            Ok(())
+        }
         [] => print_help(),
         _ => {
             println!("Invalid arguments: {:?}", &args[1..]);
             println!("\nTry again, with help: \n");
-            print_help();
+            print_help()
         }
     }
 }
 
 fn version() -> Version {
-    Version::parse(env!("CARGO_PKG_VERSION")).expect("parse CARGO_PKG_VERSION")
+    // Since `CARGO_PKG_VERSION` is provided by Cargo itself, which uses the same `semver` crate to
+    // parse version strings, the `unwrap()` below should never fail.
+    Version::parse(env!("CARGO_PKG_VERSION")).unwrap()
 }
 
-fn read_version_attribute(path: &Path) -> Version {
-    let file = fs::File::open(path).expect(&format!("Couldn't open file {}", path.display()));
+fn read_version_attribute(path: &Path) -> Result<Version> {
+    let file = fs::File::open(path).context(format!("Couldn't open file {}", path.display()))?;
     io::BufReader::new(file)
         .lines()
         .filter_map(|line| line.ok())
         .find(|line| line.trim_start().starts_with(VERSION_ATTRIBUTE_NAME))
         .and_then(|s| {
             if let Some(i) = s.find('"') {
                 if let Some(j) = s.rfind('"') {
-                    return Some(Version::parse(&s[i + 1..j]).expect("parse version attribute"));
+                    return Version::parse(&s[i + 1..j]).ok();
```

Comment:
I think we should use `Result<>` instead of `Option<>` here so that we dont have to ignore the parsing error?

Reply:
I initially considered that when working on this commit, but the problem is that the final output type of the combinator chain is `Result<Result<_, _>, _>`, which results in a somewhat clunky appearance. 😞 In this case, however, the error isn't actually being ignored as a `None` will trigger the `Valid cargo2nixVersion not found in {}` error.

I do agree that it's not as informative, though. I think I'll refactor this block in a subsequent PR to reduce the combinator nesting and deep rightward drift. I'll re-introduce the correct error reporting then.